include/wex/util.hpp

```diff
@@ -10,8 +11,65 @@ class Pinned {
 	Pinned(Pinned const&) = delete;
 	Pinned(Pinned&)		  = delete;
 
+	Pinned operator=(Pinned&&) = delete;
+	Pinned operator=(Pinned const&) = delete;
+
   public:
 	Pinned() = default;
 };
 
+#define WEX_ENSURE(cond) ((cond) ? (void(0)) : std::terminate())
+
+/// \brief An observing pointer that holds a resource and ensures that the pointer it holds
+/// is never `nullptr`.
+template <typename T>
+class NotNullPtr {
+  public:
+	NotNullPtr()				   = delete;
+	NotNullPtr(std::nullptr_t ptr) = delete;
+
+	constexpr NotNullPtr(T* ptr) : mPtr(ptr) {
```

Comment:
This API is a bit different from convention: `not_null<T*>` (not `not_null<T>`): that has two benefits: (a) can store smart pointers too (would need to move all `T`s to accommodate unique pointers), (b) clearer indication of what type the class template instantiation actually stores.

One simple way to just get (b) is to specialize `NotNull<T*>` (and keep `NotNull<T>` undefined).

Reply:
Hmm, I don't really see the problem with the current implementation seeing how it's used right now.

Reply:
The biggest "problem" _right now_ is that the type parameter is misleading: `NotNull<Foo>` actually stores and models a `Foo*`, not a `Foo`. The next biggest problem this will cause  (very soon) is when the pointer is `const`: no way to have `NotNull` model `T const*`.